api/v1alpha1/wasmextensionpolicy_types.go

```diff
@@ -0,0 +1,113 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	"google.golang.org/protobuf/types/known/anypb"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	gwapiv1a2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+const (
+	// WasmExtensionPolicy is the name of the WasmExtensionPolicy kind.
+	KindWasmExtensionPolicy = "WasmExtensionPolicy"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=envoy-gateway,shortName=sp
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Status",type=string,JSONPath=`.status.conditions[?(@.type=="Accepted")].reason`
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// WasmExtensionPolicy allows the user to configure wasm extensions for a Gateway.
+type WasmExtensionPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of WasmExtensionPolicy.
+	Spec WasmExtensionPolicySpec `json:"spec"`
+
+	// Status defines the current status of WasmExtensionPolicySpec.
+	Status gwapiv1a2.PolicyStatus `json:"status,omitempty"`
+}
+
+// WasmExtensionPolicySpec defines the desired state of WasmExtensionPolicy.
+type WasmExtensionPolicySpec struct {
+	// +kubebuilder:validation:XValidation:rule="self.group == 'gateway.networking.k8s.io'", message="this policy can only have a targetRef.group of gateway.networking.k8s.io"
+	// +kubebuilder:validation:XValidation:rule="self.kind in ['Gateway', 'HTTPRoute', 'GRPCRoute']", message="this policy can only have a targetRef.kind of Gateway/HTTPRoute/GRPCRoute"
+	// +kubebuilder:validation:XValidation:rule="!has(self.sectionName)",message="this policy does not yet support the sectionName field"
+	//
+	// TargetRef is the name of the Gateway resource this policy
+	// is being attached to.
+	// This Policy and the TargetRef MUST be in the same namespace
+	// for this Policy to have effect and be applied to the Gateway.
+	TargetRef gwapiv1a2.PolicyTargetReferenceWithSectionName `json:"targetRef"`
+
+	// Extensions is a list of Wasm extensions to be loaded by the Gateway.
+	// Order matters, as the extensions will be loaded in the order they are
+	// defined in this list.
+	//
+	// +kubebuilder:validation:MinItems=1
+	Extensions []WasmExtension `json:"extensions"`
+}
+
+// WasmExtension defines an wasm extension.
+type WasmExtension struct {
```

Comment:
Should we add `fail_open` and `capability_restriction_config` on WasmExtension ?

Reply:
Yes, they can be added later. I just want to make the first API as simple as possible.